### PR TITLE
[pushover] Improved exception handling

### PR DESCRIPTION
**Pull request by cweitkamp**
- Improved exception handling

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java [hunk 1]
```diff
@@ -279,27 +283,23 @@ public ContentProvider build() throws PushoverCommunicationException {
...
         try {
             Path tmpFile = Files.createTempFile("pushover-", ".tmp");
             return Files.write(tmpFile, data);
         } catch (IOException e) {
-            logger.debug("IOException occurred while creating temp file - skip sending message: {}",
-                    e.getLocalizedMessage(), e);
-            throw new PushoverCommunicationException(
-                    String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+            logger.debug("IOException occurred while creating temp file - skip sending message: {}", e.getMessage(), e);
+            throw new CommunicationException(String.format("Skip sending the message: %s", e.getMessage()), e);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java:292]: You could have created a `@text` entry for this message too. But you decide.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java [hunk 2]
```diff
@@ -279,27 +283,23 @@ public ContentProvider build() throws PushoverCommunicationException {
...
         try {
             body.addFilePart(MESSAGE_KEY_ATTACHMENT, path.toFile().getName(),
                     new PathContentProvider(contentType == null ? DEFAULT_CONTENT_TYPE : contentType, path), null);
         } catch (IOException e) {
-            logger.debug("IOException occurred while adding content - skip sending message: {}",
-                    e.getLocalizedMessage(), e);
-            throw new PushoverCommunicationException(
-                    String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+            logger.debug("IOException occurred while adding content - skip sending message: {}", e.getMessage(), e);
+            throw new CommunicationException(String.format("Skip sending the message: %s", e.getMessage()), e);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java:302]: Same error and this is even the same message.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 3]
```diff
@@ -101,11 +101,16 @@ public void initialize() {
...
+                if (sounds != null) {
+                    return sounds;
+                }
+            }
+        } catch (CommunicationException e) {
             // do nothing, causing exception is already logged
-        } catch (PushoverConfigurationException e) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        } catch (ConfigurationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:113]: Here, you have to use `e.getRawMessage()` not `e.getLocalizedMessage()`

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 4]
```diff
@@ -143,10 +148,10 @@ public boolean sendMessage(PushoverMessageBuilder messageBuilder) {
         if (connection != null) {
             try {
                 return connection.sendMessage(messageBuilder);
-            } catch (PushoverCommunicationException e) {
+            } catch (CommunicationException e) {
                 // do nothing, causing exception is already logged
-            } catch (PushoverConfigurationException e) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            } catch (ConfigurationException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:154]: Same here, please use `e.getRawMessage()`

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 5]
```diff
@@ -158,10 +163,10 @@ public String sendPriorityMessage(PushoverMessageBuilder messageBuilder) {
         if (connection != null) {
             try {
                 return connection.sendPriorityMessage(messageBuilder);
-            } catch (PushoverCommunicationException e) {
+            } catch (CommunicationException e) {
                 // do nothing, causing exception is already logged
-            } catch (PushoverConfigurationException e) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            } catch (ConfigurationException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:169]: Same here, please use `e.getRawMessage()`

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 6]
```diff
@@ -173,10 +178,10 @@ public boolean cancelPriorityMessage(String receipt) {
         if (connection != null) {
             try {
                 return connection.cancelPriorityMessage(receipt);
-            } catch (PushoverCommunicationException e) {
+            } catch (CommunicationException e) {
                 // do nothing, causing exception is already logged
-            } catch (PushoverConfigurationException e) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            } catch (ConfigurationException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:184]: Same here, please use `e.getRawMessage()`

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java [hunk 7]
```diff
@@ -189,8 +194,8 @@ private void asyncValidateUser() {
         try {
             connection.validateUser();
             updateStatus(ThingStatus.ONLINE);
-        } catch (PushoverCommunicationException | PushoverConfigurationException e) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+        } catch (CommunicationException | ConfigurationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
Comment (lolodomo) [inline bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:198]: Same here, please use `e.getRawMessage()`

**Review by lolodomo — CHANGES_REQUESTED** [review]
That is cool that you used these new exception classes.
I have few comments.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java [hunk 8]
```diff
@@ -152,35 +163,38 @@ private synchronized String executeRequest(HttpMethod httpMethod, String url, @N
...
+                    throw new CommunicationException(content);
             }
         } catch (ExecutionException e) {
-            logger.debug("Exception occurred during execution: {}", e.getLocalizedMessage(), e);
-            throw new PushoverCommunicationException(e.getLocalizedMessage(), e.getCause());
+            String message = e.getMessage();
+            logger.debug("Exception occurred during execution: {}", message, e);
+            throw new CommunicationException(message == null ? TEXT_OFFLINE_COMMUNICATION_ERROR : message,
+                    e.getCause());
         } catch (InterruptedException | TimeoutException e) {
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java:176]: You are talking about the `InterruptedException`? Aren't you?

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java [hunk 9]
```diff
@@ -48,62 +52,74 @@
...
-        final JsonObject sounds = json == null || !json.has("sounds") ? null : json.get("sounds").getAsJsonObject();
-
-        return sounds == null ? List.of()
-                : sounds.entrySet().stream().map(entry -> new Sound(entry.getKey(), entry.getValue().getAsString()))
+        try {
+            final String content = get(
+                    buildURL(SOUNDS_URL, Map.of(PushoverMessageBuilder.MESSAGE_KEY_TOKEN, localApikey)));
+            final JsonObject json = JsonParser.parseString(content).getAsJsonObject();
+            final JsonObject sounds = json.has(JSON_VALUE_SOUNDS) ? json.get(JSON_VALUE_SOUNDS).getAsJsonObject()
+                    : null;
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java:113]: No, both should be non null. Whereas Gson methods will throw exceptions if something tiswrong.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverAPIConnection.java:176]: Yes, of course, sorry for the mismatch.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java (hunk 4 above)
Comment (jlaur) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:154]: Hmm, this text is user-oriented (exposed in the UI), so isn't the idea exactly to use the localized text?

Comment (lolodomo) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:154]: Yes, it is and to get the localized text in updateStatus, while `@text` was used as message key in exceptions, you have to call getRawMessage(). Calling getLocalizedMessage() without any params will not return the localized text. 
See that PR: https://github.com/openhab/openhab-core/pull/2549 for the full discussion. 
Ideally, I should create a chapter to document how to use properly AbstractI18nException.

Comment (jlaur) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java:154]: > See that PR: [openhab/openhab-core#2549](https://github.com/openhab/openhab-core/pull/2549) for the full discussion.

Thanks for this explanation, very useful.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java]: Good catch. Thanks for the hint.

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java (hunk 4 above)
Comment (cweitkamp) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java]: Thanks. I kept it wrong in my mind. I probably have to fix it for OpenWeaterhMap too after https://github.com/openhab/openhab-addons/pull/11949.

**Comment by cweitkamp** [comment]
I incorporated all your comments.

**Review by lolodomo — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java (hunk 4 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/handler/PushoverAccountHandler.java]: Yes, I can confirm you have to fix OWN binding.

**Pull request merged**